File: app/src/main/java/ca/appvelopers/mcgillmobile/model/Course.java
/*
 * Copyright 2014-2016 Appvelopers
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ca.appvelopers.mcgillmobile.model;

import com.guerinet.utils.DateUtils;

import org.threeten.bp.DayOfWeek;
import org.threeten.bp.LocalDate;
import org.threeten.bp.LocalTime;

import java.io.Serializable;
import java.util.List;

import timber.log.Timber;

/**
 * A course in the user's schedule or one that a user can register for
 * @author Quang Dao
 * @author Julien Guerinet
 * @since 1.0.0
 */
public class Course implements Serializable {
    private static final long serialVersionUID = 1L;
    /**
     * The term this class is for
     */
    protected Term term;
    /**
     * The course's 4-letter subject (ex: MATH)
     */
    protected final String subject;
    /**
     * The course's number (ex: 263)
     */
    protected final String number;
    /**
     * The course title
     */
    protected final String title;
    /**
     * The course CRN number
     */
    protected final int crn;
    /**
     * The course section (ex: 001)
     */
    protected final String section;
    /**
     * The course's start time
     */
    protected final LocalTime startTime;
    /**
     * The course's end time
     */
    protected final LocalTime endTime;
    /**
     * The days this course is on
     */
    protected final List<DayOfWeek> days;
    /**
     * The course type (ex: lecture, tutorial...)
     */
    protected final String type;
    /**
     * The course location (generally building and room number)
     */
    protected final String location;
    /**
     * The course's instructor's name
     */
    protected final String instructor;
    /**
     * The number of credits for this course
     */
    protected final double credits;
    /**
     * The course start date
     */
    protected final LocalDate startDate;
    /**
     * The course end date
     */
    protected final LocalDate endDate;

    /**
     * Constructor used for the user's already registered classes
     *
     * @param subject    The course subject
     * @param number     The course number
     * @param title      The course title
     * @param crn        The course CRN
     * @param section    The course section
     * @param startTime  The course's ending time
     * @param endTime    The course's starting time
     * @param days       The days this course is on
     * @param type       The course type
     * @param location   The course location
     * @param instructor The course instructor
     * @param credits    The number of credits
     * @param startDate  The starting date for this course
     * @param endDate    The ending date for this course
     */
    public Course(String subject, String number, String title, int crn, String section,
            LocalTime startTime, LocalTime endTime, List<DayOfWeek> days, String type,
            String location, String instructor, double credits, LocalDate startDate,
            LocalDate endDate) {
        this.subject = subject;
        this.number = number;
        this.title = title;
        this.crn = crn;
        this.section = section;
        this.startTime = startTime;
        this.endTime = endTime;
        this.days = days;
        this.type = type;
        this.location = location;
        this.instructor = instructor;
        this.credits = credits;
        this.startDate = startDate;
        this.endDate = endDate;
    }

	/* GETTERS */

    /**
     * @return The course term
     */
    public Term getTerm() {
        return term;
    }

    /**
     * @return The course code
     */
    public String getCode() {
        return subject + " " + number;
    }

    /**
     * @return The course subject
     */
    public String getSubject() {
        return subject;
    }

    /**
     * @return The course number
     */
    public String getNumber() {
        return number;
    }

    /**
     * @return The course title
     */
    public String getTitle() {
        return title;
    }

    /**
     * @return The course CRN
     */
    public int getCRN() {
        return crn;
    }

    /**
     * @return The course section
     */
    public String getSection() {
        return section;
    }

    /**
     * @return The course start time
     */
    public LocalTime getStartTime() {
        return startTime;
    }


    /**
     * @return The course end time
     */
    public LocalTime getEndTime() {
        return endTime;
    }

    /**
     * @return The days this course is on
     */
    public List<DayOfWeek> getDays() {
        return days;
    }

    /**
     * @return The course type
     */
    public String getType() {
        return type;
    }

    /**
     * @return The course's location
     */
    public String getLocation() {
        return location;
    }

    /**
     * @return The course's instructor
     */
    public String getInstructor() {
        return instructor;
    }

    /**
     * @return The course credits
     */
    public double getCredits() {
        return credits;
    }

    /**
     * @return The starting date
     */
    public LocalDate getStartDate() {
        return startDate;
    }

    /**
     * @return The ending date
     */
    public LocalDate getEndDate() {
        return endDate;
    }

    /* SETTERS */

    /**
     * @param term Course {@link Term}
     */
    public void setTerm(Term term) {
        this.term = term;
    }

    /* HELPERS */

    /**
     * Returns the start time of the course, rounded off to the nearest half hour
     *
     * @return The course rounded start time
     */
    public LocalTime getRoundedStartTime() {
        //Check if the start time is already a half hour increment
        if (startTime.getMinute() == 0 || startTime.getMinute() == 30) {
            return startTime;
        }

        //If not, remove 5 minutes to the start time to get round numbers
        //  (McGill start times are always 5 minutes after the nearest half hour)
        return startTime.minusMinutes(5);
    }

    /**
     * Returns the end time of the course, rounded off the the nearest half hour
     *
     * @return The course rounded end time
     */
    public LocalTime getRoundedEndTime() {
        //Check if the end time is already a half hour increment
        if (endTime.getMinute() == 0 || endTime.getMinute() == 30) {
            return endTime;
        }

        //If not, add 5 minutes to the end time to get round numbers
        //  (McGill end times are always 5 minutes before the nearest half hour
        return endTime.plusMinutes(5);
    }

    /**
     * @param date The given date
     * @return True if the course if for the given date, false otherwise
     */
    public boolean isForDate(LocalDate date) {
        //Check if the date is within the date range and the course is offered on that day
        return !date.isBefore(startDate) && !date.isAfter(endDate) &&
                days.contains(date.getDayOfWeek());
    }

    /**
     * @return The course times in String format
     */
    public String getTimeString() {
        //No time associated, therefore no time string
        if (startTime.getHour() == 0 && startTime.getMinute() == 0) {
            Timber.i("No time associated when getting String");
            return "";
        }

        return DateUtils.getShortTimeString(startTime) + " - " +
                DateUtils.getShortTimeString(endTime);
    }

    /**
     * @return The course dates in String format
     */
    public String getDateString() {
        return DateUtils.getMediumDateString(startDate) + " - " +
                DateUtils.getMediumDateString(endDate);
    }

    /**
     * 2 ClassItems are equal if they have the same CRN and are for the same term
     *
     * @param object The course to check
     * @return True if they are equal, false otherwise
     */
    @Override
    public boolean equals(Object object) {
        if (!(object instanceof Course)) {
            return false;
        }

        Course course = (Course) object;

        //Check if they have the same season, year, and CRN
        return crn == course.getCRN() && term.equals(course.getTerm());
    }
}

File: app/src/main/java/model/Course.kt
/*
 * Copyright 2014-2022 Julien Guerinet
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.guerinet.mymartlet.model

import androidx.room.Entity
import androidx.room.PrimaryKey
import com.guerinet.suitcase.date.android.extensions.getMediumDateString
import com.guerinet.suitcase.util.Utils
import kotlinx.datetime.DayOfWeek
import kotlinx.datetime.LocalDate
import timber.log.Timber

/**
 * A course in the user's schedule or one that a user can register for
 * @author Julien Guerinet
 * @author Quang Dao
 * @since 1.0.0
 *
 * @property term Term this class is for
 * @property subject Course's 4-letter subject (ex: MATH)
 * @property number Course's number (ex: 263)
 * @property title Course title
 * @property crn Course CRN number
 * @property section Course section (ex: 001)
 * @property startTime Course's start time
 * @property endTime Course's end time
 * @property days Days this course is on
 * @property type Course type (ex: lecture, tutorial...)
 * @property location Course location (generally building and room number)
 * @property instructor Course's instructor's name
 * @property credits Number of credits for this course
 * @property startDate Course start date
 * @property endDate Course end date
 * @property id Unique Id for this course
 */
@Entity
open class Course(
    var term: Term,
    val subject: String,
    val number: String,
    val title: String,
    val crn: Int,
    val section: String,
    val startTime: LocalTime,
    val endTime: LocalTime,
    val days: List<DayOfWeek>,
    val type: String,
    val location: String,
    val instructor: String,
    val credits: Double,
    val startDate: LocalDate,
    val endDate: LocalDate,
    @PrimaryKey var id: String = Utils.uuid()
) {

    /** Course code */
    val code: String
        get() = "$subject $number"

    /** Start time of the course, rounded off to the nearest half hour */
    val roundedStartTime: LocalTime
        // Check if the start time is already a half hour increment
        //  If not, remove 5 minutes to the start time to get round numbers
        //  (McGill start times are always 5 minutes after the nearest half hour)
        get() = if (startTime.minute == 0 || startTime.minute == 30) {
            startTime
        } else startTime.minusMinutes(5)

    /** End time of the course, rounded off the the nearest half hour */
    val roundedEndTime: LocalTime
        // Check if the end time is already a half hour increment
        // If not, add 5 minutes to the end time to get round numbers
        //  (McGill end times are always 5 minutes before the nearest half hour
        get() = if (endTime.minute == 0 || endTime.minute == 30) {
            endTime
        } else endTime.plusMinutes(5)

    /** Course times in String format */
    val timeString: String
        get() {
            if (startTime.hour == 0 && startTime.minute == 0) {
                Timber.i("No time associated when getting String")
                return ""
            }
            return "${startTime.getShortTimeString()} - ${endTime.getShortTimeString()}"
        }

    /** Course dates in String format */
    val dateString: String
        get() = "${startDate.getMediumDateString()} - ${endDate.getMediumDateString()}"

    /**
     * Returns true if the course is for the [date], false otherwise
     *  If the date is within the date range and that the course is offered on that day
     */
    fun isForDate(date: LocalDate): Boolean =
        date in startDate..endDate && days.contains(date.dayOfWeek)
}
